### PR TITLE
340 per pcsp marginal likelihoods

### LINKED ISSUE #340: Output per-PCSP marginal likelihoods as part of GP optimization

### PR DESCRIPTION
**Pull request by hrnasif**
## Description

Updates optimization to a per-pcsp notion of convergence. Branch length estimation convergence is now based on average change in branch lengths. 

Other updates include:
- Changing the default tolerance for Brent optimization from 6 to 10.
- Adding functionality to export per-pcsp log likelihoods.
- Adding/updating gradient-based methods for branch length optimization. These include Newton Optimization and log-space gradient ascent. Default is still set at Brent Optimization without gradients.
- Various functions used in prior analyses, such as obtaining per pcsp likelihood surfaces, fidgeting single branch lengths to evaluate how different methods would re-optimize, and the option to track "intermediate" branch length values during branch length estimation. 

Closes #340 

## Tests

Test to verify correct derivative and second derivative calculations
Test comparing Newton optimization and Brent's method, verifying that they arrive at similar solutions for a small dataset

## Checklist:

* [x] Code follows our detailed [contribution guidelines](CONTRIBUTING.md)
* [x] `clang-format` has been run
* [x] TODOs have been eliminated from the code
* [x] Comments are up to date, document intent, and there are no commented-out code blocks
* [x] Issue branch has been squashed and rebased on main branch
* [x] GitHub CI build on PR branch completed successfully

### CODE REVIEW

**Review by matsen — COMMENTED** [review]
I had a snippet of time and just added a few questions in header files.

File: src/optimization.hpp [hunk 1]
```diff
@@ -55,32 +60,171 @@ std::pair<T, T> BrentMinimize(F f, T min, T max, int significant_digits,
...
+      delta = golden * delta2;
+    }
+
+    // ** Update current position:
+    u = (fabs(delta) >= fract1)
+            ? T(x + delta)
+            : (delta > 0 ? T(x + fabs(fract1)) : T(x - fabs(fract1)));
+    fu = f(u).first;
+
+    // Considering update using gradient descent:
```
Comment (davidrich27) [inline src/optimization.hpp:221]: Is this section at the end the extent of the differences between gradient and non-gradient Brent?  If so, you could add a template parameter to Brent to use gradients or not that could be resolved at compile time, and remove some redundant code.

**Review by davidrich27 — COMMENTED** [review]
I'll go over this again, but this is what stood out to me on first pass.  Looks good!

File: src/optimization.hpp (hunk 1 above)
Comment (hrnasif) [reply src/optimization.hpp]: Updated based on your suggestion, please take a look

File: src/gp_engine.cpp [hunk 2]
```diff
@@ -590,23 +704,94 @@ void GPEngine::BrentOptimization(const GPOperations::OptimizeBranchLength& op) {
...
-      relative_tolerance_for_optimization_, step_size_for_optimization_,
+  const auto branch_length = Optimization::GradientAscent(
+      log_likelihood_and_derivative, branch_lengths_(op.gpcsp_),
+      significant_digits_for_optimization_, step_size_for_optimization_,
       min_log_branch_length_, max_iter_for_optimization_);
-  std::ignore = log_likelihood;
+  branch_lengths_(op.gpcsp_) = branch_length;
+}
+
+// Fixes log-space branch length gradient ascient optimization
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.cpp]: This was in reference to an automatic fail on the gradient ascent method. It's been removed

File: src/gp_engine.hpp [hunk 3]
```diff
@@ -95,6 +109,8 @@ class GPEngine {
   EigenVectorXd GetBranchLengths(const size_t start, const size_t length) const;
   // Get Branch Lengths from temporary space.
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.hpp:109]: @davidrich27 can you comment on this? It looks like this is a method to get branch lengths in a specific block

File: src/gp_instance.hpp [hunk 4]
```diff
@@ -116,5 +150,15 @@ class GPInstance {
   GPDAG dag_;
   static constexpr size_t plv_count_per_node_ = 6;
 
+  size_t gpcsp_count_ = dag_.EdgeCountWithLeafSubsplits();
+  bool use_gradients_;
+  GPEngine::OptimizationMethod optimization_method_;
+
+  // For optimization tracking
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_instance.hpp]: Added more details to these

File: src/gp_doctest.cpp [hunk 5]
```diff
@@ -28,18 +28,20 @@ enum HelloGPCSP { jupiter, mars, saturn, venus, rootsplit, root };
...
+  inst.UseGradientOptimization(use_gradients);
   inst.MakeEngine();
   return inst;
 }
 
 // Our tree is (see check below)
 // (jupiter:0.113,(mars:0.15,saturn:0.1)venus:0.22):0.;
 // You can see a helpful diagram at
-// https://github.com/phylovi/bito/issues/349#issuecomment-898672399
+// https://github.com/phylovi/libsbn/issues/213#issuecomment-624195267
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_doctest.cpp]: Nope! Updated

File: src/gp_engine.hpp [hunk 6]
```diff
@@ -58,6 +58,16 @@ class GPEngine {
 
   // ** GPOperations
 
+  // Options for optimization method.
+  enum class OptimizationMethod {
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.hpp:62]: We set two default values, a default gradient and default non-gradient value. These are specified gp_engine.cpp

File: src/gp_engine.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/gp_engine.hpp:109]: Yes, these are these are the branch lengths for grafted edges.

File: src/optimization.hpp (hunk 1 above)
Comment (davidrich27) [reply src/optimization.hpp]: Yep, that's what I had in mind.  I think it looks good!

**Comment by davidrich27** [comment]
@hrnasif You should remove the DS_Store files from the commit.  That is just metadata from your MacOS.

File: src/gp_engine.cpp [hunk 7]
```diff
@@ -571,17 +641,58 @@ double GPEngine::LogRescalingFor(size_t plv_idx) {
...
+void GPEngine::Optimization(const GPOperations::OptimizeBranchLength& op) {
+  return Optimization(op, optimization_method_);
+}
+
+void GPEngine::Optimization(const GPOperations::OptimizeBranchLength& op,
+                            std::optional<GPEngine::OptimizationMethod> os) {
+  Assert(os.has_value(),
+         "GPEngine::Optimization(): Optimization method has not been set.");
+
+  switch (os.value()) {
```
Comment (davidrich27) [inline src/gp_engine.cpp:657]: I think making the optimization method `std::optional` was from me, before we added the optimization method assignment to the constructor and removed the default method from enum.  But now I think we can probably just remove `std::optional` from these functions and the `optimization_method_` variable.

File: src/pybito.cpp [hunk 8]
```diff
@@ -625,7 +643,17 @@ PYBIND11_MODULE(bito, m) {
            "Estimate the SBN parameters based on current branch lengths.")
       .def("estimate_branch_lengths", &GPInstance::EstimateBranchLengths,
            "Estimate branch lengths for the GPInstance.", py::arg("tol"),
-           py::arg("max_iter"), py::arg("quiet") = false)
+           py::arg("max_iter"), py::arg("quiet") = false,
+           py::arg("intermediate") = false)
```
Comment (davidrich27) [inline src/pybito.cpp:647]: Change `intermediate` to match argument name in `GPInstance`.

**Review by davidrich27 — COMMENTED** [review]
Just some minor suggestions here.  Everything else looks good!  Also, see my comment above about DS_Store files.

File: src/gp_engine.cpp [hunk 9]
```diff
@@ -147,6 +154,13 @@ void GPEngine::GrowGPCSPs(const size_t new_gpcsp_count,
   // Initialize new work space.
   for (size_t i = old_gpcsp_count; i < GetPaddedGPCSPCount(); i++) {
     branch_lengths_[i] = default_branch_length_;
+
+    // Initialize branch length difference, with dag root node equal 0
+    if (i == old_gpcsp_count) {
+      branch_length_differences_[i] = 0;
+    } else {
+      branch_length_differences_[i] = default_branch_length_;
+    }
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.cpp]: So it only occurs for one index and only on initialization. I don't think we'd want to change this loop in the general case when we are growing the pcsp count. I tried to rewrite this block to specifically address this in my latest commit.

File: src/gp_engine.hpp [hunk 10]
```diff
@@ -241,13 +268,14 @@ class GPEngine {
   // space).
   static constexpr double max_log_branch_length_ = 1.1;
   // Precision used for checking convergence of branch length optimization.
-  int significant_digits_for_optimization_ = 6;
-  //
-  double relative_tolerance_for_optimization_ = 1e-2;
-  // Step size used for gradient-based branch length optimization.
+  int significant_digits_for_optimization_ = 10;
```
Comment (matsen) [inline src/gp_engine.hpp:271]: Wasn't there something surprising about the way this was set up? Like it was some number of significant digits in the binary representation or ...? A comment would help.

File: src/optimization.hpp [hunk 11]
```diff
@@ -5,16 +5,17 @@
...
 
 #include "sugar.hpp"
 
 namespace Optimization {
 
-// Copied from https://www.boost.org/doc/libs/1_73_0/boost/math/tools/minima.hpp
-template <class F, class T>
-std::pair<T, T> BrentMinimize(F f, T min, T max, int significant_digits,
-                              size_t max_iter) {
+// Adapted from https://www.boost.org/doc/libs/1_73_0/boost/math/tools/minima.hpp
```
Comment (matsen) [inline src/optimization.hpp:15]: Did you come up with the Brent + gradient yourself? If not, is there some reference?

File: src/pybito.cpp [hunk 12]
```diff
@@ -625,7 +643,17 @@ PYBIND11_MODULE(bito, m) {
...
-           py::arg("max_iter"), py::arg("quiet") = false)
+           py::arg("max_iter"), py::arg("quiet") = false,
+           py::arg("track_intermediate_iterations") = false)
+      .def("get_perpcsp_llh_surface", &GPInstance::GetPerGPCSPLogLikelihoodSurfaces,
+           "Scan the likelihood surface for the pcsps in the GPInstance.",
+           py::arg("steps"), py::arg("scale_min"), py::arg("scale_max"))
+      .def("perturb_and_track_optimization_values",
+           &GPInstance::PerturbAndTrackValuesFromOptimization,
+           "Reinitiate optimization, perturbing one branch length at a time,  and "
+           "track branch length and per pcsp likelihoods/")
```
Comment (matsen) [inline src/pybito.cpp:654]: Does trailing slash indicate something is missing?

**Review by matsen — CHANGES_REQUESTED** [review]
This looks really nice, Hassan! 

It would be great if you could include a short paragraph as a comment somewhere that would describe your experience with the optimizers, basically just saying that Brent is the way to go... with gradients? I'm not sure.

File: src/pybito.cpp (hunk 12 above)
Comment (hrnasif) [reply src/pybito.cpp:654]: Nope, fixed.

File: src/gp_engine.cpp [hunk 13]
```diff
@@ -571,17 +632,50 @@ double GPEngine::LogRescalingFor(size_t plv_idx) {
...
+
 void GPEngine::BrentOptimization(const GPOperations::OptimizeBranchLength& op) {
   auto negative_log_likelihood = [this, &op](double log_branch_length) {
-    SetTransitionMatrixToHaveBranchLength(exp(log_branch_length));
+    SetTransitionAndDerivativeMatricesToHaveBranchLength(exp(log_branch_length));
     PreparePerPatternLogLikelihoodsForGPCSP(op.rootward_, op.leafward_);
     return -per_pattern_log_likelihoods_.dot(site_pattern_weights_);
   };
+
+  if (branch_length_differences_(op.gpcsp_) < branch_length_difference_threshold_) {
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.cpp:667]: Done

File: src/gp_engine.cpp [hunk 14]
```diff
@@ -590,23 +684,93 @@ void GPEngine::BrentOptimization(const GPOperations::OptimizeBranchLength& op) {
...
 }
 
-void GPEngine::GradientAscentOptimization(
+void GPEngine::BrentOptimizationWithGradients(
     const GPOperations::OptimizeBranchLength& op) {
-  // See issue #278:
-  Failwith(
-      "Hasn't been tested after the change to log-space branch length optimization.");
-  auto log_likelihood_and_derivative = [this, &op](double log_branch_length) {
+  auto negative_log_likelihood = [this, &op](double log_branch_length) {
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.cpp:693]: Right, updated with accurate name

File: src/gp_engine.cpp [hunk 15]
```diff
@@ -590,23 +684,93 @@ void GPEngine::BrentOptimization(const GPOperations::OptimizeBranchLength& op) {
...
+    double branch_length = exp(log_branch_length);
+    branch_lengths_(op.gpcsp_) = branch_length;
+    auto [log_likelihood, log_likelihood_derivative] =
+        this->LogLikelihoodAndDerivative(op);
+    return std::make_pair(-log_likelihood, -branch_length * log_likelihood_derivative);
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.cpp:698]: We're performing minimization in log space, so the second term is the derivative of the negative likelihood with respect to log(branch_length)

File: src/gp_engine.cpp [hunk 16]
```diff
@@ -571,17 +632,50 @@ double GPEngine::LogRescalingFor(size_t plv_idx) {
...
+}
+
+void GPEngine::SetOptimizationTolerance(int optim_tol) {
+  significant_digits_for_optimization_ = optim_tol;
+}
+
 void GPEngine::BrentOptimization(const GPOperations::OptimizeBranchLength& op) {
   auto negative_log_likelihood = [this, &op](double log_branch_length) {
-    SetTransitionMatrixToHaveBranchLength(exp(log_branch_length));
+    SetTransitionAndDerivativeMatricesToHaveBranchLength(exp(log_branch_length));
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.cpp:662]: Not needed; an artifact from trying to add gradients to Brent before templating. Updated back.

File: src/gp_engine.cpp [hunk 17]
```diff
@@ -571,17 +632,50 @@ double GPEngine::LogRescalingFor(size_t plv_idx) {
...
+    case OptimizationMethod::LogSpaceGradientAscentOptimization:
+      return LogSpaceGradientAscentOptimization(op);
+    case OptimizationMethod::NewtonOptimization:
+      return NewtonOptimization(op);
+    default:
+      Failwith("GPEngine::Optimization(): Invalid OptimizationMethod given.");
+  }
+}
+
+void GPEngine::SetOptimizationTolerance(int optim_tol) {
```
[the remark this answers is not in the document]
Comment (hrnasif) [reply src/gp_engine.cpp:656]: Done

File: src/gp_engine.hpp (hunk 10 above)
Comment (hrnasif) [reply src/gp_engine.hpp:271]: Added some text describing this

File: src/optimization.hpp (hunk 11 above)
Comment (hrnasif) [reply src/optimization.hpp:15]: I modified the algorithm myself. I think there is some version of Brent + gradients out there (replacing the secant method with Newton's method in the root-finding version of Brent), but that's not what we're doing here.

**Review by matsen — APPROVED** [review]
So glad to see this PR landing, @hrnasif ! Congrats.

Note that there's just a little merge conflict that needs fixing before merging. After that, you or @davidrich27 can squash and merge.

**Pull request merged**